backend/icpy/core/framework_compatibility.py

```
@@ -343,6 +345,35 @@ def _attachment_to_image_part(self, att: Dict[str, Any]) -> Optional[Dict[str, A
+                try:
+                    import os
+                    from pathlib import Path
+                    ws_root = os.environ.get('WORKSPACE_ROOT')
+                    abs_path = Path(preferred_fs_path).resolve()
+                    # Only allow reading from within WORKSPACE_ROOT when set
+                    if ws_root:
+                        ws_root_p = Path(ws_root).resolve()
+                        abs_path.relative_to(ws_root_p)
+                    # If WORKSPACE_ROOT is not set, as a safety fallback disallow absolute paths outside cwd/workspace
+                    elif not abs_path.is_file():
+                        raise ValueError('Invalid path')
```

Ruff (0.13.1)

Abstract `raise` to an inner function

(TRY301)